### DISCUSSION TITLE
Error generating static pages - DYNAMIC_SERVER_USAGE

### DISCUSSION THREAD

**Discussion by GasparAdragna** [amannn/next-intl#1062] [Q&A]
Hi!
I currently have:
```typescript
export function generateStaticParams() {
  return locales.map((locale) => ({ locale }));
}
``` 
and 

```typescript
unstable_setRequestLocale(locale);
``` 

on my `[locale]/layout.tsx` file.

But when i try to build im getting the following error on most of my pages:

`
description: "Route /en/account couldn't be rendered statically because it used headers. See more info here: https://nextjs.org/docs/messages/dynamic-server-error",
    digest: 'DYNAMIC_SERVER_USAGE'
`

I'm NOT using `unstable_setRequestLocale` on the pages that it's giving error since this comment https://github.com/amannn/next-intl/issues/663#issuecomment-1852536760 said i do not need it in `"use client"` pages.
Also these pages have the information of the user and should be dynamic.

My questions are: Should i ignore this errors and move on (the build seems to run correctly although the errors on the console)? Am i doing the right approach here? Should i remove the `generateStaticParams` and go full dynamic?

Thanks!

PS:  i use this approach for the titles of the pages: i create a `layout.tsx` on each page. Is this the correct approach as well?

```typescript
import { getTranslations } from "next-intl/server";

import Page from "./page";

export const generateMetadata = async ({
  params: { locale },
}: {
  params: { locale: "en" | "es" };
}) => {
  const t = await getTranslations({ locale, namespace: "pages" });

  return {
    title: t("account.myBankAccounts.title"),
  };
};

export default Page;

```

**Comment by amannn** [comment]
It might be because Next.js doesn't allow dynamic routes within layouts that use `generateStaticParams`: https://github.com/vercel/next.js/issues/44712. I haven't personally experimented with that though, also not sure if something has changed on the Next.js side since that issue was created.

**Options to solve this:**
1. Don't use static generation (i.e. remove `generateStaticParams`)
1. Make sure all pages can be statically generated
2. Move `generateStaticParams` from the layout to relevant pages

**Reply by e-roy** [reply]
I didn't realize that `generateStaticParams` could be used on the page level.  I didn't notice it in the docs, maybe I missed it, but if it's not there this would be great to add.   Gives much better control for which pages should be static :)

**Reply by amannn** [reply]
That's a good point, thanks for mentioning this! Added this in https://github.com/amannn/next-intl/commit/12fd5c558c691867b2f65b909e6b68ca41ee62a9.

**Comment by el-j** [comment]
hi @amannn i came here and i see what you suggest but i do not understand how to solve it. i am in need for "output: 'export'" 
i have read this section and the mentioned links: [next-intl-docs](https://next-intl-docs.vercel.app/docs/routing/middleware#usage-without-middleware-static-export)

```
"next": "v15.0.0-canary.6",
"next-intl": "^3.14.1",
```
<img width="1011" alt="image" src="https://github.com/amannn/next-intl/assets/2795534/be02d375-18fb-451d-b98b-a2839eeffae7">

how do i have to configure it to work with the mentioned `getRequestConfig` function?
<img width="541" alt="image" src="https://github.com/amannn/next-intl/assets/2795534/2fd8fe30-7f08-4d9f-a26e-4d34f335f28b">

i tried to move the mentioned generateStaticParms from `layout` to `page` but i still have this error, with the main config. 
is there any way i can achieve static rendering (output: export) again? was working with these versions:
```
"next": "13.5.4",
"next-intl": "^3.0.0-beta.19"
```

**Reply by amannn** [reply]
Have you seen this? https://next-intl-docs.vercel.app/docs/getting-started/app-router/with-i18n-routing#static-rendering

**Reply by zhiga90** [reply]
> Have you seen this? https://next-intl-docs.vercel.app/docs/getting-started/app-router/with-i18n-routing#static-rendering

@amannn 
Sorry, my english is bad, use a translator.

I took a new clean project starting with npx create-next-app@latest and did everything according to your instructions from the link above, but still got a bunch of errors for `npm run build` like `Usage of next-intl APIs in Server Components currently opts into dynamic rendering.` or `Route /ru couldn't be rendered statically because it used headers.` .

can be viewed here:
`https://github.com/zhiga90/next-intl-test`

Please show me where I am wrong.

**Comment by portmaler** [comment]
I had the same issue and I solved it by adding   unstable_setRequestLocale(locale);  thi si my code: 
```
import { notFound } from 'next/navigation';
import { getRequestConfig, unstable_setRequestLocale } from 'next-intl/server';

const locales = ['en', 'fr', 'ar'];

export default getRequestConfig(async ({ locale }) => {
  // Validate locale for static rendering using unstable_setRequestLocale
  if (!locales.includes(locale as any)) notFound();

  // Use the unstable_setRequestLocale to enable static generation
  unstable_setRequestLocale(locale); // Set the locale for static generation

  return {
    messages: (await import(`../messages/${locale}.json`)).default
  };
});
```

**Reply by MahmoudKSultan** [reply]
locale binding right now is undefined and as docs said in the new version, you gotta use requestLocale instead

**Comment by ayush-goyal** [comment]
I had the same issue. I fixed it by adding `export const dynamic = "force-dynamic";` to the top of the `page.tsx` component. This allows you to take advantage of the `generateStaticParams()` and still re-render dynamic pages for each user.

**Comment by MahmoudKSultan** [comment]
Have you figured it out?
